src/Bicep.Core/Analyzers/Linter/Rules/NoHardcodedEnvironmentUrlsRule.cs

```diff
@@ -10,89 +10,121 @@
         }
 
         public override string FormatMessage(params object[] values)
             => string.Format("{0} Found this disallowed host: \"{1}\"", this.Description, values.First());
 
-        /// <summary>
-        /// Note that this regex requires the match to be at a subdomain boundary
-        /// so that there is not a partial match "mid-word". The subdomain must
-        /// be proceeded by a text break (., or slash, or space) or at the beginning
-        /// of a line.
-        /// </summary>
-        /// <returns></returns>
-        public Regex CreateDisallowedHostRegex() =>
-            new Regex(string.Join('|', this.DisallowedHosts!.Value.Select(h => $@"(?<=\.|'|{{|}}|\s|^|/){Regex.Escape(h)}")),
-                        RegexOptions.Compiled | RegexOptions.IgnoreCase);
-
-        public Regex CreateExcludedHostsRegex() =>
-            new Regex(string.Join('|', this.ExcludedHosts!.Value.Select(h => $@"(?<=\.|'|{{|}}|\s|^|/){Regex.Escape(h)}")),
-                        RegexOptions.Compiled | RegexOptions.IgnoreCase);
-
         public override IEnumerable<IDiagnostic> AnalyzeInternal(SemanticModel model)
         {
             if (HasHosts)
             {
-                var visitor = new Visitor(this.hostRegex, this.MinimumHostLength, this.excludedRegex);
+                var visitor = new Visitor(this.disallowedHosts ?? ImmutableArray<string>.Empty, this.MinimumHostLength, this.excludedHosts ?? ImmutableArray<string>.Empty);
                 visitor.Visit(model.SourceFile.ProgramSyntax);
+
                 return visitor.DisallowedHostSpans.Select(entry => CreateDiagnosticForSpan(entry.Key, entry.Value));
             }
+
             return Enumerable.Empty<IDiagnostic>();
         }
 
-        private class Visitor : SyntaxVisitor
+        private sealed class Visitor : SyntaxVisitor
         {
             public readonly Dictionary<TextSpan, string> DisallowedHostSpans = new Dictionary<TextSpan, string>();
-            private readonly Regex hostRegex;
+            private readonly ImmutableArray<string> disallowedHosts;
             private readonly int minHostLen;
-            private readonly Regex exclusionRegex;
+            private readonly ImmutableArray<string> excludedHosts;
 
-            public Visitor(Regex disallowedRegex, int minHostLen, Regex exclusionRegex)
+            public Visitor(ImmutableArray<string> disallowedHosts, int minHostLen, ImmutableArray<string> excludedHosts)
             {
-                this.hostRegex = disallowedRegex;
+                this.disallowedHosts = disallowedHosts;
                 this.minHostLen = minHostLen;
-                this.exclusionRegex = exclusionRegex;
+                this.excludedHosts = excludedHosts;
+            }
+
+            public static IEnumerable<(TextSpan RelativeSpan, string Value)> FindHostnameMatches(string hostname, string srcText)
+            {
+                bool isExactDomainMatch(int index)
```

Comment:
>isExactDomainMatch [](http://example.com/codeflow?start=21&length=18)

This isn't really checking for any match. It's looking before and after the match to ensure the chars aren't alphanumeric.

Should we rename to something like `CheckNoSurroundingAlphaNumericChars` or `CheckNoLeadingAndNoTrailingAlphaNumericChars`?